lib/ansible/context.py

```diff
@@ -0,0 +1,53 @@
+# Copyright: (c) 2018, Toshio Kuratomi <tkuratomi@ansible.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+"""
+Context of the running Ansible.
+
+In the future we *may* create Context objects to allow running multiple Ansible plays in parallel
+with different contexts but that is currently out of scope as the Ansible library is just for
+running the ansible command line tools.
+
+These APIs are still in flux so do not use them unless you are willing to update them with every Ansible release
+"""
+
+from ansible.arguments.context_objects import CLIArgs, GlobalCLIArgs
+
+
+# Note: this is not the singleton version.  That is only created once the program has actually
+# parsed the args
+CLIARGS = CLIArgs({})
+
+
+class _Context:
```

Comment:
What are the plans for this class?  Upon further inspection, I don't see it being used.

Reply:
tldr; Yeah, it could be removed.

More info:

@nitzmahone's criticisms of the initial idea of global singleton's was that it prevents us from having different arguments in different threads/workers down the road.  Having a context would fix that (Basically reinstating the old way of passing in the cli args from the toplevel into the functions which do the work but being expandable to more than just cli args).  When I was struggling with the unittests I started implementing a Context object so that I could pass it in for unittests to workaround the single-immutable aspect of the CLIARGS.  After adopting your idea to mess with the Singleton's private data instead, though, I didn't need to continue down the Context path.

The Context object did point out to me the need for separate GlobalCLIArgs and CliArgs objects which I've now implemented but we don't need to keep the Context code as proof of that.  I should probably add more information to the CliArgs class to justify its existence if I remove Context from here, though.

Reply:
Removed in latest version